src/processes/poisson.jl

```diff
@@ -19,61 +24,94 @@ Base.union(p₁::PoissonProcess{<:Real}, p₂::PoissonProcess{<:Function}) = Poi
 
 Base.union(p₁::PoissonProcess{<:Function}, p₂::PoissonProcess{<:Real}) = PoissonProcess(x -> p₁.λ(x) + p₂.λ)
 
+Base.union(p₁::PoissonProcess{<:AbstractVector}, p₂::PoissonProcess{<:AbstractVector}) = PoissonProcess(x -> p₁.λ + p₂.λ)
+
 ishomogeneous(p::PoissonProcess{<:Real}) = true
 ishomogeneous(p::PoissonProcess{<:Function}) = false
-
-default_sampling_algorithm(::PoissonProcess, ::Geometry) = DiscretizedSampling()
-default_sampling_algorithm(::PoissonProcess{<:Real}, ::Box) = ProductSampling()
+ishomogeneous(p::PoissonProcess{<:Vector}) = false
+
+default_sampling_algorithm(::PoissonProcess) = DiscretizedSampling()
+default_sampling_algorithm(::PoissonProcess{<:Function}) = ThinnedSampling()
+
+default_lambda_max(p, g, algo::ThinnedSampling{T}) where {T<:Real} = algo.param
+
+function default_lambda_max(p, g, algo::ThinnedSampling{T}) where {T<:Dims}
+  # compute intensity on vertices
+  box = boundingbox(g)
+  if isnothing(algo.param)
+    m = CartesianGrid(box.min, box.max)
+  else
+    m = CartesianGrid(box.min, box.max, dims = algo.param)
+  end
+  v = vertices(m)
+
+  # obtain λmax
+  λvec = p.λ.(v)
+  λmax = maximum(λvec) + 0.05 * (maximum(λvec) - minimum(λvec))
+end
 
 #------------------
 # HOMOGENEOUS CASE
 #------------------
 
-function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Real}, g::Geometry, ::DiscretizedSampling)
+function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Real}, g, ::DiscretizedSampling)
   # simulate number of points
   λ = p.λ
   V = measure(g)
   n = rand(rng, Poisson(λ * V))
 
-  # simulate homogeneous process
-  pts = sample(g, HomogeneousSampling(n))
+  if iszero(n)
+    # PointSet{embeddim(g), coordtype(g)}([])
+    nothing
+  else
+    # simulate homogeneous process
+    points = sample(g, HomogeneousSampling(n))
 
-  # return point pattern
-  PointSet(collect(pts))
+    # return point pattern
+    PointSet(points)
+  end
 end
 
-function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Real}, b::Box, ::ProductSampling)
-  # region configuration
-  lo, up = coordinates.(extrema(b))
+#--------------------
+# INHOMOGENEOUS CASE
+#--------------------
 
+function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Vector}, d::Domain, algo::DiscretizedSampling)
   # simulate number of points
   λ = p.λ
-  V = measure(b)
-  n = rand(rng, Poisson(λ * V))
-
-  # product of uniform distributions
-  U = product_distribution([Uniform(lo[i], up[i]) for i in 1:embeddim(b)])
-
-  # return point pattern
-  PointSet(rand(rng, U, n))
+  V = measure.(d)
+  n = rand(rng, Poisson(sum(λ .* V)))
+
+  if iszero(n)
+    nothing
+  else
+    # sample elements with weights proportial to expected number of points
+    w = WeightedSampling(n, λ .* V, replace = true)
+
+    # within each element sample a single point
+    sampler = HomogeneousSampling(1)
+    points = (first(sample(rng, e, sampler)) for e in sample(rng, d, w))
+
+    # return point pattern
+    PointSet(points)
```

Comment:
This implementation is exactly equal to the Meshes.jl implementation for general Domain:

https://github.com/JuliaGeometry/Meshes.jl/blob/5f64dea3cf27cd555fec90d9eda6498bb6cc1c69/src/sampling/homogeneous.jl#L15-L26

If we could use that implementation instead, that would be ideal. Maybe we could simply scale the geometries in a given Domain according to the lambda vector and call the Meshes.jl method? Something along these lines would reuse the code we have there and would also be more intuitive to read.

Reply:
Maybe we could use the Stretch transform on each geometry of the mesh in a lazy fashion, i.e. stretching boxes should be zero effort. We may need to add specializations in our Stretch implementation for specific geometries though: https://github.com/JuliaGeometry/Meshes.jl/blob/5f64dea3cf27cd555fec90d9eda6498bb6cc1c69/src/transforms/stretch.jl#L5-L26